File: src/main/java/org/apache/prepbuddy/cluster/TextFacets.java
package org.apache.prepbuddy.cluster;

import org.apache.prepbuddy.exceptions.SystemException;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.function.Function2;
import scala.Tuple2;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * TextFacets contains the cardinal values of the column with their
 * number of occurrence in that column.
 */
public class TextFacets implements Serializable {
    private final JavaPairRDD<String, Integer> facets;

    public TextFacets(JavaPairRDD<String, Integer> facets) {
        this.facets = facets;
    }

    public long count() {
        return facets.count();
    }

    public List<Tuple2> highest() {
        return getPeakListFor(new Function2<Integer, Integer, Boolean>() {
            @Override
            public Boolean call(Integer currentTupleValue, Integer peakTupleValue) throws Exception {
                return currentTupleValue > peakTupleValue;
            }
        });
    }

    public List<Tuple2> lowest() throws Exception {
        return getPeakListFor(new Function2<Integer, Integer, Boolean>() {
            @Override
            public Boolean call(Integer currentTupleValue, Integer peakTupleValue) throws Exception {
                return currentTupleValue < peakTupleValue;
            }
        });
    }

    private List<Tuple2> getPeakListFor(Function2<Integer, Integer, Boolean> compareFunction) {
        List<Tuple2<String, Integer>> allTuple = facets.collect();
        ArrayList<Tuple2> list = new ArrayList<>();
        try {
            Tuple2<String, Integer> peakTuple = allTuple.get(0);
            list.add(peakTuple);
            for (Tuple2<String, Integer> tuple : allTuple) {
                if (compareFunction.call(tuple._2(), peakTuple._2())) {
                    peakTuple = tuple;
                    list.clear();
                    list.add(peakTuple);
                }

                if (tuple._2().equals(peakTuple._2()) && !(tuple.equals(peakTuple))) {
                    list.add(tuple);
                }

            }
        } catch (Exception e) {
            throw new SystemException(e);
        }
        return list;
    }


    public JavaPairRDD<String, Integer> rdd() {
        return facets;
    }

    public List<Tuple2> getFacetsBetween(int lowerBound, int upperBound) {
        List<Tuple2<String, Integer>> allTuple = facets.collect();
        ArrayList<Tuple2> list = new ArrayList<>();

        for (Tuple2<String, Integer> tuple : allTuple) {
            Integer currentTupleValue = tuple._2();

            if (isInRange(currentTupleValue, lowerBound, upperBound))
                list.add(tuple);
        }
        return list;
    }

    private boolean isInRange(Integer currentTupleValue, int minimum, int maximum) {
        return currentTupleValue >= minimum && currentTupleValue <= maximum;
    }


    public List<String> cardinalValues() {
        List<Tuple2<String, Integer>> tuples = facets.collect();
        ArrayList<String> keys = new ArrayList<>();
        for (Tuple2<String, Integer> tuple : tuples) {
            keys.add(tuple._1());
        }
        return keys;
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/clusterers/TextFacets.scala
package org.apache.datacommons.prepbuddy.clusterers

import org.apache.datacommons.prepbuddy.utils.Range
import org.apache.spark.rdd.RDD

import scala.collection.mutable.ListBuffer

/**
  * TextFacets is a collection of unique strings and the number of times the string appears in a column
  */
class TextFacets(facets: RDD[(String, Int)]) {
    private val tuples: Array[(String, Int)] = facets.collect()

    def getFacetsBetween(lowerBound: Int, upperBound: Int): Array[(String, Int)] = {
        tuples.filter(tuple => new Range(lowerBound, upperBound).contains(tuple._2))
    }

    def lowest: Array[(String, Int)] = getPeakListFor(_ < _)

    def highest: Array[(String, Int)] = getPeakListFor(_ > _)

    private def getPeakListFor(compareFunction: (Int, Int) => Boolean): Array[(String, Int)] = {
        val option: Option[(String, Int)] = tuples.find(!_._1.isEmpty)
        var peakTuple = option.head
        var facetsCount: ListBuffer[(String, Int)] = ListBuffer(peakTuple)
        tuples.view.filter(!_._1.trim.isEmpty).foreach((tuple) => {
            if (compareFunction(tuple._2, peakTuple._2)) {
                peakTuple = tuple
                facetsCount = ListBuffer(peakTuple)
            }
            if ((tuple._2 == peakTuple._2) && (tuple != peakTuple)) {
                facetsCount += tuple
            }
        })
        facetsCount.toArray
    }

    def count: Long = facets.count

    def cardinalValues: Array[String] = tuples.map(_._1)

    def rdd: RDD[(String, Int)] = facets
}
